## Optimize the Model


In [104]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf

#  Import and read the charity_data.csv.
import pandas as pd 
application_df = pd.read_csv("charity_data.csv")
application_df.head()

EIN                                      NAME  ... ASK_AMT IS_SUCCESSFUL
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB  ...    5000             1
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR  ...  108590             1
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS  ...    5000             0
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION  ...    6692             1
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT  ...  142590             1

[5 rows x 12 columns]

In [105]:
application_df.nunique()

EIN                       34299
NAME                      19568
APPLICATION_TYPE             17
AFFILIATION                   6
CLASSIFICATION               71
USE_CASE                      5
ORGANIZATION                  4
STATUS                        2
INCOME_AMT                    9
SPECIAL_CONSIDERATIONS        2
ASK_AMT                    8747
IS_SUCCESSFUL                 2
dtype: int64

In [107]:
application_df['STATUS'].value_counts()

1    34294
0        5
Name: STATUS, dtype: int64

In [88]:
application_df['SPECIAL_CONSIDERATIONS'].value_counts()

N    34272
Y       27
Name: SPECIAL_CONSIDERATIONS, dtype: int64

Status and Special Considerations can be dropped as majority of data are in one bucket i.e. have one value.

In [108]:
# Drop the non-beneficial EIN column
application_df=application_df.drop(columns=['EIN','SPECIAL_CONSIDERATIONS','STATUS'])
application_df.head()

NAME  ... IS_SUCCESSFUL
0              BLUE KNIGHTS MOTORCYCLE CLUB  ...             1
1    AMERICAN CHESAPEAKE CLUB CHARITABLE TR  ...             1
2        ST CLOUD PROFESSIONAL FIREFIGHTERS  ...             0
3            SOUTHSIDE ATHLETIC ASSOCIATION  ...             1
4  GENETIC RESEARCH INSTITUTE OF THE DESERT  ...             1

[5 rows x 9 columns]

In [109]:
# display the number of unique values

application_df.nunique()

NAME                19568
APPLICATION_TYPE       17
AFFILIATION             6
CLASSIFICATION         71
USE_CASE                5
ORGANIZATION            4
INCOME_AMT              9
ASK_AMT              8747
IS_SUCCESSFUL           2
dtype: int64

In [110]:
from pandas.core.indexes.multi import names_compat
# Look at NAME value counts for binning
name_counts= pd.DataFrame(application_df.NAME.value_counts().reset_index())
name_counts.columns=['name','count']
#  How many name counts are greater than 5?
name_counts[name_counts['count']>5]

name  count
0                             PARENT BOOSTER USA INC   1260
1                                      TOPS CLUB INC    765
2                 UNITED STATES BOWLING CONGRESS INC    700
3                        WASHINGTON STATE UNIVERSITY    492
4    AMATEUR ATHLETIC UNION OF THE UNITED STATES INC    408
..                                               ...    ...
349            OLD OAK CLIFF CONSERVATION LEAGUE INC      6
350           AMERICAN NEPHROLOGY NURSES ASSOCIATION      6
351        HUMBLE ISD EDUCATIONAL SUPPORT GROUPS INC      6
352              PROFESSIONAL LOADMASTER ASSOCIATION      6
353                                         CBMC INC      6

[354 rows x 2 columns]

In [111]:
# Determine which values to replace if counts are less than or equal to 5.
names_to_replace = [n for n in name_counts[name_counts['count']<=5]['name']]

# Replace in dataframe
for name in names_to_replace:
    application_df['NAME'] = application_df['NAME'].replace(name,"Other")
    
# Check to make sure binning was successful
application_df['NAME'].value_counts()

Other                                           20043
PARENT BOOSTER USA INC                           1260
TOPS CLUB INC                                     765
UNITED STATES BOWLING CONGRESS INC                700
WASHINGTON STATE UNIVERSITY                       492
                                                ...  
HABITAT FOR HUMANITY INTERNATIONAL                  6
DAMAGE PREVENTION COUNCIL OF TEXAS                  6
FLEET RESERVE ASSOCIATION                           6
HUGH OBRIAN YOUTH LEADERSHIP                        6
INTERNATIONAL CONGRESS OF CHURCHES MINISTERS        6
Name: NAME, Length: 355, dtype: int64

In [112]:
# Look at APPLICATION_TYPE value counts for binning
application_df.APPLICATION_TYPE.value_counts()

T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: APPLICATION_TYPE, dtype: int64

In [113]:
# Determine which values to replace if application counts are less than 500
application_types_count=pd.DataFrame(application_df['APPLICATION_TYPE'].value_counts().reset_index())
application_types_count.columns=['app_type','count']
application_types_to_replace=[t for t in application_types_count[application_types_count['count']<500]['app_type']]

# Replace in dataframe
for app in application_types_to_replace:
    application_df['APPLICATION_TYPE'] = application_df['APPLICATION_TYPE'].replace(app,"Other")

# Check to make sure binning was successful
application_df['APPLICATION_TYPE'].value_counts()

T3       27037
T4        1542
T6        1216
T5        1173
T19       1065
T8         737
T7         725
T10        528
Other      276
Name: APPLICATION_TYPE, dtype: int64

In [114]:
# Look at CLASSIFICATION value counts for binning
application_df['CLASSIFICATION'].value_counts()

C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C4120        1
C8210        1
C2561        1
C4500        1
C2150        1
Name: CLASSIFICATION, Length: 71, dtype: int64

In [115]:
# Determine which values to replace if counts are less than 1000
class_types_count=pd.DataFrame(application_df['CLASSIFICATION'].value_counts().reset_index())
class_types_count.columns=['class_type','count']
classifications_to_replace=[t for t in class_types_count[class_types_count['count']<1000]['class_type']]
# Replace in dataframe
for cls in classifications_to_replace:
    application_df['CLASSIFICATION'] = application_df['CLASSIFICATION'].replace(cls,"Other")
    
# Check to make sure binning was successful
application_df['CLASSIFICATION'].value_counts()

C1000    17326
C2000     6074
C1200     4837
Other     2261
C3000     1918
C2100     1883
Name: CLASSIFICATION, dtype: int64

In [116]:
# Generate our categorical variable lists
[c for c in application_df.columns if application_df[c].dtype == 'object']


['NAME',
 'APPLICATION_TYPE',
 'AFFILIATION',
 'CLASSIFICATION',
 'USE_CASE',
 'ORGANIZATION',
 'INCOME_AMT']

In [117]:
# perform one-hot encoding on the data
application_df_dummies=pd.get_dummies(application_df)
application_df_dummies.head()


ASK_AMT  IS_SUCCESSFUL  ...  INCOME_AMT_50M+  INCOME_AMT_5M-10M
0     5000              1  ...                0                  0
1   108590              1  ...                0                  0
2     5000              0  ...                0                  0
3     6692              1  ...                0                  0
4   142590              1  ...                0                  0

[5 rows x 396 columns]

In [118]:
# Split our preprocessed data into our features and target arrays

X = application_df_dummies.drop(columns=['IS_SUCCESSFUL']).values
y = application_df_dummies['IS_SUCCESSFUL'].values

# Split the preprocessed data into a training and testing dataset
X_train,X_test,y_train,y_test=train_test_split(X,y)

In [119]:
# Create a StandardScaler instances

scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)


In [136]:
len(X_train_scaled[0])

395

In [137]:
# Define the model - deep neural net
number_input_features = len(X_train_scaled[0])
hidden_nodes_layer1 =  100
hidden_nodes_layer2 = 30
hidden_nodes_layer3 = 10

nn = tf.keras.models.Sequential()

# First hidden layer
nn.add(
    tf.keras.layers.Dense(units=hidden_nodes_layer1, input_dim=number_input_features, activation="relu")
)

# Second hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer2, activation="sigmoid"))

# Third hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer3, activation="sigmoid"))

# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn.summary()

Model: "sequential_5"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_20 (Dense)            (None, 100)               39600     
                                                                 
 dense_21 (Dense)            (None, 30)                3030      
                                                                 
 dense_22 (Dense)            (None, 10)                310       
                                                                 
 dense_23 (Dense)            (None, 1)                 11        
                                                                 
Total params: 42,951
Trainable params: 42,951
Non-trainable params: 0
_________________________________________________________________


In [138]:
# Compile the model
nn.compile(loss='binary_crossentropy',optimizer="adam",metrics=['accuracy'])

In [139]:
# Train the model
fit_model = nn.fit(X_train_scaled,y_train,epochs=100)

Epoch 1/100
804/804 [==============================] - 3s 3ms/step - loss: 0.4918 - accuracy: 0.7717
Epoch 2/100
804/804 [==============================] - 3s 3ms/step - loss: 0.4434 - accuracy: 0.7943
Epoch 3/100
804/804 [==============================] - 2s 3ms/step - loss: 0.4351 - accuracy: 0.7959
Epoch 4/100
804/804 [==============================] - 2s 3ms/step - loss: 0.4318 - accuracy: 0.7968
Epoch 5/100
804/804 [==============================] - 2s 3ms/step - loss: 0.4289 - accuracy: 0.7971
Epoch 6/100
804/804 [==============================] - 2s 3ms/step - loss: 0.4267 - accuracy: 0.7976
Epoch 7/100
804/804 [==============================] - 2s 3ms/step - loss: 0.4248 - accuracy: 0.7982
Epoch 8/100
804/804 [==============================] - 2s 2ms/step - loss: 0.4237 - accuracy: 0.7994
Epoch 9/100
804/804 [==============================] - 2s 2ms/step - loss: 0.4227 - accuracy: 0.7993
Epoch 10/100
804/804 [==============================] - 2s 3ms/step - loss: 0.4214 - accura

In [140]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 1s - loss: 0.4565 - accuracy: 0.7924 - 524ms/epoch - 2ms/step
Loss: 0.45647814869880676, Accuracy: 0.7924198508262634


In [141]:
# Export our model to HDF5 file
nn.save("AlphabetSoupCharity_Optimization.h5")

USE A RANDOM FOREST CLASSIFIER ON THE DATASET

In [126]:
from sklearn.metrics import accuracy_score
from sklearn.ensemble import RandomForestClassifier

In [127]:
from sklearn.model_selection import GridSearchCV
from sklearn.ensemble import RandomForestRegressor

In [128]:
param_grid = {
  
    'n_estimators': [100, 150, 200, 300]
}

In [142]:
random_forest_model = RandomForestClassifier()
grid_search = GridSearchCV(estimator = random_forest_model , param_grid = param_grid, cv = 3)

In [143]:
grid_search.fit(X_train_scaled,y_train)

GridSearchCV(cv=3, estimator=RandomForestClassifier(),
             param_grid={'n_estimators': [100, 150, 200, 300]})

In [144]:
print(grid_search.best_params_,grid_search.best_estimator_,grid_search.best_score_)

{'n_estimators': 100} RandomForestClassifier() 0.7732082824132352


In [145]:
# Create a random forest classifier.
rf_model=RandomForestClassifier(n_estimators=100,random_state=82)

# Fitting the model
rf_model=rf_model.fit(X_train_scaled,y_train)

# Evaluate the model
y_pred=rf_model.predict(X_test_scaled)
print(f"Random forest model accuracy: {accuracy_score(y_test,y_pred)*100:.2f}")

Random forest model accuracy: 77.82


In [134]:
application_df.head()

NAME APPLICATION_TYPE  ... ASK_AMT IS_SUCCESSFUL
0  BLUE KNIGHTS MOTORCYCLE CLUB              T10  ...    5000             1
1                         Other               T3  ...  108590             1
2                         Other               T5  ...    5000             0
3                         Other               T3  ...    6692             1
4                         Other               T3  ...  142590             1

[5 rows x 9 columns]

**OverView:-** 

Create an algorithm to predict whether or not applicants funding will be successful. It is desired to have 75% accuracy

**Results:-**

**Data Preprocessing**

*   Target -> Is Successful i.e Was money used effectively.
*   Features Considered for Model -> Name,Application Type,Affiliated Sector of industry,Govt. Organization Classification,Use case for funding,Organization Type,Income Classification,Funding Amount Requested

*   Features removed from input data -> EIN,Status,Special Consideration for application

**Model**

*   There are 5 layers, 1 input, 3 Hidden and 1 Output layer. Input layer has 395 neurons (i.e lenght of X_SCALED[0], 1st hidden Layer has 100, 2nd hidden layer has 30 and 3rd hidden layer has 10
*   Yes Target model performance is achieved.Accurracy of the model us 79%.

*   To improve model performance deleted low cardinality features (like status,Special consideration) and included name which has high cardinality. Also moved around the neurons in each layer till target accuracy was exceeded.

**Summary**

Overall our model predicts the success of applicants funding with 79% accuracy. 
We can run the model using more epochs, use Kmeans clustering to see data clusters and use that to define neurons at each level to improve the model further.






